### PR TITLE
[ShapProp] Refactor TensorMeta to use qparams for more general quantization support

### PR DESCRIPTION
**Pull request by jerryzh168**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #65287
* __->__ #64787

Summary:
This PR refactors TensorMeta to use qparams dictionary which will be scalable for
all future quantization schemes, this will be a bc-breaking change for TensorMeta,
but since bc-guarantee has not been checkedin for 1.9 we can still do it. This PR should
be landed before 1.10 branch cut.

Test Plan:
python torch/fx/experimental/fx2trt/example/quantized_resnet_test.py

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D30879848](https://our.internmc.facebook.com/intern/diff/D30879848)

### CODE REVIEW

File: torch/fx/passes/shape_prop.py [hunk 1]
```diff
@@ -21,6 +21,9 @@ class TensorMetadata(NamedTuple):
     qscheme : Optional[torch.qscheme]
     q_scale : Optional[float]
     q_zero_point : Optional[int]
+    q_per_channel_scales : Optional[torch.Tensor]
```
Comment (jerryzh168) [inline torch/fx/passes/shape_prop.py:24]: Hi @jamesr66a, it looks like TensorMeta needs to be backward compatible, how can we add new things to it?

Comment (jamesr66a) [reply torch/fx/passes/shape_prop.py:24]: Add your class member with a default value that's assigned in the body of the class: https://docs.python.org/3/library/typing.html#typing.NamedTuple

After you do that, `--accept` test_class_member_back_compat and tag me for review

**Comment by jerryzh168** [comment]
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30879848).

**Comment by jerryzh168** [comment]
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30879848).

File: torch/fx/experimental/fx_acc/acc_ops.py [hunk 2]
```diff
@@ -389,30 +388,26 @@ def min_single_tensor_input(*, input):
...
-    ],
-    kwargs_to_move_to_acc_out_ty=[
-        ("dtype", "dtype"),
-        ("scale", "q_scale"),
-        ("zero_point", "q_zero_point"),
-    ],
-)
 @register_acc_op
-def quantize_per_tensor(*, input, acc_out_ty=None):
+def quantize_per_tensor(*, input, qparams=None):
```
Comment (842974287) [inline torch/fx/experimental/fx_acc/acc_ops.py:392]: We might still want to use `acc_out_ty` here to be consistent with other quantized ops. Instead we can change how `kwargs_to_move_to_acc_out_ty` works so it would create `qparams` correctly.

File: torch/fx/experimental/fx2trt/converters/acc_ops_converters.py [hunk 3]
```diff
@@ -1367,23 +1408,35 @@ def acc_ops_dequantize(network, target, args, kwargs, name):
...
+        if q_zero_point != 0:
+            raise RuntimeError(f"Only support zero_point == 0, get {q_zero_point}")
+    elif qscheme == torch.per_channel_affine:
+        q_scale = qparams["scale"]
+        q_zero_point = qparams["zero_point"]
+        q_axis = qparams["axis"]
+        scale_shape = q_scale.shape
+        if not torch.equal(q_zero_point, torch.zeros(q_zero_point.shape, dtype=q_zero_point.dtype)):
+            raise RuntimeError(f"Only support zero_point == 0, get {q_zero_point}")
+
```
[the remark this answers is not in the document]
Comment (jerryzh168) [reply torch/fx/experimental/fx2trt/converters/acc_ops_converters.py:1426]: yeah sure

File: torch/fx/experimental/fx_acc/acc_ops.py [hunk 4]
```diff
@@ -359,6 +358,51 @@ def matmul(*, input, other):
     return torch.matmul(**locals())
 
 
+@register_custom_acc_mapper_fn(
```
Comment (842974287) [inline torch/fx/experimental/fx_acc/acc_ops.py:361]: Remove? Seems like something wrong happened during rebasing.

File: torch/fx/experimental/fx_acc/acc_normalizer.py [hunk 5]
```diff
@@ -60,6 +60,9 @@ class NormalizationInfo(NamedTuple):
     custom_mapping_fn: Optional[Callable]
     kwargs_to_move_to_acc_out_ty: Optional[Optional[List[Tuple[str, str]]]]
     needs_shapes_for_normalization: bool
+    # qscheme of Tensor, used to get different fields for quantization
+    # parameters
+    qscheme: Optional[torch.qscheme]
```
Comment (842974287) [inline torch/fx/experimental/fx_acc/acc_normalizer.py:65]: `qscheme` doesn't feel like a good fit in `NormalizationInfo`. Why do we need this? I saw it mostly just hard coded for some ops.

File: torch/fx/experimental/fx_acc/acc_ops.py (hunk 4 above)
Comment (jerryzh168) [reply torch/fx/experimental/fx_acc/acc_ops.py:361]: oh right

File: torch/fx/experimental/fx_acc/acc_normalizer.py [hunk 6]
```diff
@@ -243,26 +235,38 @@ def move_kwargs_to_acc_out_ty(
...
-            qparams[tmd_field_name] = new_kwargs[orig_kwarg_name]
+    for kwarg_replacement_tuple in normalization_info.kwargs_to_move_to_acc_out_ty:
+        if len(kwarg_replacement_tuple) == 2:
+            orig_kwarg_name, tmd_field_name, is_constant = *kwarg_replacement_tuple, False
+        else:
+            assert len(kwarg_replacement_tuple) == 3
+            orig_kwarg_name, tmd_field_name, is_constant = kwarg_replacement_tuple
+        if is_constant:
+            # when the arg is constant, the fields will be name and value
+            tmd_dict[orig_kwarg_name] = tmd_field_name
```
Comment (842974287) [inline torch/fx/experimental/fx_acc/acc_normalizer.py:246]: Can we make the tuple (constant, tmd_field_name, is_constant) so that here we can do `tmd_dict[tmd_field_name) = orig_kwarg_name`?

**Review by 842974287 — APPROVED** [review]
Looks good to me! Just one more comment and cc: @jfix71 to take another look at the changes on `kwargs_to_move_to_acc_out_ty`.

File: torch/fx/experimental/fx_acc/acc_normalizer.py (hunk 6 above)
Comment (jerryzh168) [reply torch/fx/experimental/fx_acc/acc_normalizer.py:246]: oh OK

**Review by jamesr66a — COMMENTED** [review]
Can you separate out the ShapeProp changes into a separate PR? That will make it easier to keep track of the origin of BC-breaking changes

**Comment by jerryzh168** [comment]
> Can you separate out the ShapeProp changes into a separate PR? That will make it easier to keep track of the origin of BC-breaking changes

sure, moved out the quantize_per_channel changes to a separate PR, please take a look again

**Comment by jerryzh168** [comment]
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30879848).

**Comment by jerryzh168** [comment]
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30879848).

**Comment by jerryzh168** [comment]
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30879848).

File: torch/fx/passes/shape_prop.py [hunk 7]
```diff
@@ -45,20 +43,27 @@ def _extract_tensor_metadata(result : torch.Tensor) -> TensorMetadata:
...
     if is_quantized:
-        qscheme = result.qscheme()
-
-        if qscheme in {torch.per_tensor_affine, torch.per_tensor_symmetric}:
-            q_scale = result.q_scale()
-            q_zero_point = result.q_zero_point()
-
+        qparams = {}
+        qparams["qscheme"] = result.qscheme()
+        qparams["dtype"] = result.dtype  # type: ignore[assignment]
```
Comment (jamesr66a) [inline torch/fx/passes/shape_prop.py:56]: What error are these `# type: ignore` comments suppressing? Can we type qparams as `Dict[str, Any]` and make this go away?

**Review by jamesr66a — APPROVED** [review]
This looks fine to me. I suggest also requesting review from the owners of the other components that are being changed (fx2trt, ACC tracer, graph drawer) and make sure those changes are OK

**Comment by jerryzh168** [comment]
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30879848).

**Comment by jerryzh168** [comment]
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30879848).

**Comment by jerryzh168** [comment]
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30879848).

**Comment by jerryzh168** [comment]
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30879848).

**Comment by jerryzh168** [comment]
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30879848).

**Comment by jerryzh168** [comment]
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30879848).

**Comment by jerryzh168** [comment]
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30879848).

**Comment by jerryzh168** [comment]
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30879848).
